src/library/scala/collection/IterableOnce.scala

```diff
@@ -558,10 +558,19 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
   /** Apply `f` to each element for its side effects
     *  Note: [U] parameter needed to help scalac's type inference.
     */
-  def foreach[U](f: A => U): Unit = {
-    val it = iterator
-    while(it.hasNext) f(it.next())
-  }
+  def foreach[U](f: A => U): Unit =
+    this match {
```

Comment:
This might be a minor detail, but may I ask why you decided to “effectively override” the method in `IterableOnce` as opposed to “actually overriding” the method in `IndexedSeqOps`?

My personal preference is to use pattern-matching style when working with a closed type hierarchy (sealed traits), and to use `override` style when working with an open type hierarchy. I prefer not to put much “knowledge” about `IterableOnce` subtypes (such as `IndexedSeq`) within `IterableOnce` itself.

I wonder if it can have an impact on performance, also.

Reply:
because `IndexedSeqOps` is a trait, and overriding it there generates a new mixing forwarder. I don't believe this is forward binary compatible (MiMa at least says it isn't)

Reply:
I'm also happy to put a private method on `IndexedSeqOps` and call that from `IterableOnceOps`, to cut down on method size if that's deemed important